### PR TITLE
executor: copy row data instead of refereeing chunk.Row in some window functions.

### LINKED ISSUE #11614: window function rank() sometimes lead to tidb-server panic
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
**1) prepare data**
same mock data set like issue: https://github.com/pingcap/tidb/issues/11612 
**2) do query**

SELECT      p.id,     p.title,     p.seq,     p.m_id,     p.val,     rank() over(PARTITION BY p.m_id ORDER BY p.seqDESC) AS rank_num FROM     testtable p   WHERE p.deleted = 0      AND p.online = 1      AND p.m_id IN (1000,1001,1002,1003,11004);

2. What did you expect to see?
rank result works well.

3. What did you see instead?

ERROR 2013 (HY000): Lost connection to MySQL server during query error returned.

**tidb-server log:**

[2019/08/05 15:12:05.408 +08:00] [ERROR] [conn.go:599] ["connection running loop panic"] [conn=6661] [lastCmd="SELECT      p.id,  p.title,     p.seq,     p.m_id,     p.val,     rank() over(PARTITION BY p.m_id ORDER BY p.seq DESC) AS rank_num FROM     testtable p   WHERE p.deleted = 0      AND p.online = 1      AND p.m_id IN (1000,1001,1002,1003,1004)"] [err="\"index out of range\""] [stack="goroutine 62192060 [running]:\ngithub.com/pingcap/tidb/server.(*clientConn).Run.func1(0x2250900, 0xc00af21a70, 0xc00e38b520)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:597 +0xee\npanic(0x1d85fe0, 0x3198a70)\n\t/usr/local/go/src/runtime/panic.go:522 +0x1b5\ngithub.com/pingcap/tidb/server.(*clientConn).writeResultset.func1(0x0, 0x22626c0, 0xc00d720190, 0xc001553778, 0x2250900, 0xc00af21a70, 0xc00e38b520)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1255 +0x59c\npanic(0x1d85fe0, 0x3198a70)\n\t/usr/local/go/src/runtime/panic.go:522 +0x1b5\ngithub.com/pingcap/tidb/util/chunk.(*column).isNull(...)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/util/chunk/column.go:71\ngithub.com/pingcap/tidb/util/chunk.Row.IsNull(...)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/util/chunk/row.go:223\ngithub.com/pingcap/tidb/util/chunk.cmpInt64(0xc01a693650, 0x1f, 0x4, 0xc00ec882d0, 0x0, 0x4, 0xc02567e414)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/util/chunk/compare.go:69 +0x171\ngithub.com/pingcap/tidb/executor/aggfuncs.(*rowComparer).compareRows(0xc001b67468, 0xc01a693650, 0x1f, 0xc00ec882d0, 0x0, 0xc02567e400)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/aggfuncs/func_rank.go:94 +0xd9\ngithub.com/pingcap/tidb/executor/aggfuncs.(*rank).AppendFinalResult2Chunk(0xc001b67440, 0x2289c80, 0xc01503e690, 0xc01a693530, 0xc01a6935f0, 0x15, 0x400)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/aggfuncs/func_rank.go:59 +0x99\ngithub.com/pingcap/tidb/executor.(*aggWindowProcessor).appendResult2Chunk(0xc01a693560, 0x2289c80, 0xc01503e690, 0xc022e96400, 0x1, 0x20, 0xc01a6935f0, 0x2, 0x1e824e0, 0x0, ...)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:213 +0xca\ngithub.com/pingcap/tidb/executor.(*WindowExec).appendResult2Chunk(0xc0124222d0, 0xc01a6935f0, 0x0, 0x0)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:155 +0xc1\ngithub.com/pingcap/tidb/executor.(*WindowExec).Next(0xc0124222d0, 0x2250900, 0xc00af21a70, 0xc01a6935f0, 0x0, 0x0)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:66 +0x1db\ngithub.com/pingcap/tidb/executor.Next(0x2250900, 0xc00af21a70, 0x2257880, 0xc0124222d0, 0xc01a6935f0, 0x1d, 0x20000015535c0)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/executor.go:190 +0xbd\ngithub.com/pingcap/tidb/executor.(*recordSet).Next(0xc00d720140, 0x2250900, 0xc00af21a70, 0xc01a6935f0, 0x0, 0x0)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/adapter.go:109 +0xb0\ngithub.com/pingcap/tidb/server.(*tidbResultSet).Next(0xc00d720190, 0x2250900, 0xc00af21a70, 0xc01a6935f0, 0x0, 0x0)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/driver_tidb.go:365 +0x51\ngithub.com/pingcap/tidb/server.(*clientConn).writeChunks(0xc00e38b520, 0x2250900, 0xc00af21a70, 0x22626c0, 0xc00d720190, 0xc000003700, 0x2250900, 0xc00af21a70)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1302 +0x319\ngithub.com/pingcap/tidb/server.(*clientConn).writeResultset(0xc00e38b520,0x2250900, 0xc00af21a70, 0x22626c0, 0xc00d720190, 0xc000000300, 0x0, 0x0, 0x0)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1268 +0x1a1\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc00e38b520, 0x2250900, 0xc00af21a70, 0xc00a46f001, 0xf4, 0x0, 0x0)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1185 +0x212\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch(0xc00e38b520, 0x2250900, 0xc00af21a70, 0xc00a46f001, 0xf5,0xf5, 0x0, 0x0)\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:895 +0x5c6\ngithub.com/pingcap/tidb/server.(*clientConn).Run(0xc00e38b520, 0x2250900, 0xc00af21a70)\n\t/home/jenkins/workspa"] [stack="github.com/pingcap/tidb/server.(*clientConn).Run.func1\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:599\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:522\ngithub.com/pingcap/tidb/server.(*clientConn).writeResultset.func1\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1255\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:522\nruntime.panicindex\n\t/usr/local/go/src/runtime/panic.go:44\ngithub.com/pingcap/tidb/util/chunk.(*column).isNull\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/util/chunk/column.go:71\ngithub.com/pingcap/tidb/util/chunk.Row.IsNull\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/util/chunk/row.go:223\ngithub.com/pingcap/tidb/util/chunk.cmpInt64\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/util/chunk/compare.go:69\ngithub.com/pingcap/tidb/executor/aggfuncs.(*rowComparer).compareRows\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/aggfuncs/func_rank.go:94\ngithub.com/pingcap/tidb/executor/aggfuncs.(*rank).AppendFinalResult2Chunk\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/aggfuncs/func_rank.go:59\ngithub.com/pingcap/tidb/executor.(*aggWindowProcessor).appendResult2Chunk\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:213\ngithub.com/pingcap/tidb/executor.(*WindowExec).appendResult2Chunk\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:155\ngithub.com/pingcap/tidb/executor.(*WindowExec).Next\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:66\ngithub.com/pingcap/tidb/executor.Next\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/executor.go:190\ngithub.com/pingcap/tidb/executor.(*recordSet).Next\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/adapter.go:109\ngithub.com/pingcap/tidb/server.(*tidbResultSet).Next\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/driver_tidb.go:365\ngithub.com/pingcap/tidb/server.(*clientConn).writeChunks\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1302\ngithub.com/pingcap/tidb/server.(*clientConn).writeResultset\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1268\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuery\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:1185\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:895\ngithub.com/pingcap/tidb/server.(*clientConn).Run\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/conn.go:650\ngithub.com/pingcap/tidb/server.(*Server).onConn\n\t/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/server/server.go:440"]

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

mysql> select version();
+--------------------+
| version()          |
+--------------------+
| 5.7.25-TiDB-v3.0.1 |
+--------------------+
1 row in set (0.00 sec)

mysql> show variables like '%sql_mode%';
+---------------+--------------------------------------------+
| Variable_name | Value                                      |
+---------------+--------------------------------------------+
| sql_mode      | STRICT_TRANS_TABLES,NO_ENGINE_SUBSTITUTION |
+---------------+--------------------------------------------+
1 row in set (0.01 sec)

### LINKED ISSUE #11626: using window function "first_value" get "index out of range" runtime error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

SELECT first_value(p.id) over(PARTITION BY p.m_id ORDER BY p.seq DESC) AS first_id FROM     testtable p   WHERE p.deleted = 0      AND p.online = 1      AND p.m_id IN (1977409, 1861420);

2. What did you expect to see?

return correct first_id  column value (first id in each group)

3. What did you see instead?

for some given m_id set, the sql run into an error:
ERROR 1105 (HY000): runtime error: index out of range

**tidb-server log:** 
 [err="runtime error: index out of range
github.com/pingcap/tidb/executor.recoveryHashAgg
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/aggregate.go:322
github.com/pingcap/tidb/executor.(*HashAggExec).fetchChildData.func1
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/aggregate.go:542
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:522
runtime.panicindex
	/usr/local/go/src/runtime/panic.go:44
github.com/pingcap/tidb/util/chunk.(*column).isNull
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/util/chunk/column.go:71
github.com/pingcap/tidb/util/chunk.Row.IsNull
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/util/chunk/row.go:223
github.com/pingcap/tidb/expression.(*Column).EvalInt
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/expression/column.go:221
github.com/pingcap/tidb/executor/aggfuncs.(*value4Int).evaluateRow
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/aggfuncs/func_value.go:41
github.com/pingcap/tidb/executor/aggfuncs.(*firstValue).UpdatePartialResult
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/aggfuncs/func_value.go:245
github.com/pingcap/tidb/executor.(*rangeFrameWindowProcessor).appendResult2Chunk
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:410
github.com/pingcap/tidb/executor.(*WindowExec).appendResult2Chunk
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:155
github.com/pingcap/tidb/executor.(*WindowExec).Next
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/window.go:66
github.com/pingcap/tidb/executor.Next
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/executor.go:190
github.com/pingcap/tidb/executor.(*HashAggExec).fetchChildData
	/home/jenkins/workspace/release_tidb_3.0/go/src/github.com/pingcap/tidb/executor/aggregate.go:558
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1337"]

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
mysql> select version();
+--------------------+
| version() |
+--------------------+
| 5.7.25-TiDB-v3.0.1 |
+--------------------+
1 row in set (0.00 sec)

mysql> show variables like '%sql_mode%';
+---------------+--------------------------------------------+
| Variable_name | Value |
+---------------+--------------------------------------------+
| sql_mode | STRICT_TRANS_TABLES,NO_ENGINE_SUBSTITUTION |
+---------------+--------------------------------------------+
1 row in set (0.01 sec)

### PR DESCRIPTION
**Pull request by SunRunAway**
### What problem does this PR solve? 

Fixes #11614, and fixes #11626 .
If a `chunk.Chunk` reuse into a window function executor, it may break the internal memory reference in function `RANK` or `PERCENT_RANK`.

1. Father calls `chk := NewChunk()`
2. Father calls `windowExecutor.Next(chk)`
3. The `windowExecutor` put column results into `chk` while refereeing the `chunk.Row` into [`PartialResult`](https://github.com/pingcap/tidb/blob/f98ef22/executor/aggfuncs/aggfuncs.go#L89) of  function `RANK` or `PERCENT_RANK`.
4. Father calls `windowExecutor.Next(chk)` again to collect next batch of result and reuses the same `chk`.
5. `chunk.Row` refereced by `PartialResult` is broken when `chk.Reset` is called.

- [x] TODO: fix the other window functions which may lead this problem.

### What is changed and how it works?

Change step 3 above, copy row data instead of refereeing chunk.Row in window functions which rely on data comparing.

### Check List 

Tests 

 - Unit test

Code changes

- None

Side effects

 - Possible performance regression

* note: no func `lag` because it cannot be finished with old code.
```
name                                                           old time/op    new time/op    delta
WindowRows/(func:row_number,_ndv:10,_rows:1000)-12               65.7µs ± 9%    68.5µs ± 8%       ~     (p=0.523 n=3+3)
WindowRows/(func:row_number,_ndv:1000,_rows:1000)-12              131µs ± 2%     161µs ± 8%    +22.55%  (p=0.041 n=3+3)
WindowRows/(func:row_number,_ndv:10,_rows:100000)-12             5.98ms ± 1%    5.97ms ± 1%       ~     (p=0.784 n=3+3)
WindowRows/(func:row_number,_ndv:1000,_rows:100000)-12           6.19ms ± 2%    6.07ms ± 0%       ~     (p=0.173 n=3+3)
WindowFunctions/(func:row_number,_ndv:1000,_rows:100000)-12      6.07ms ± 2%    6.06ms ± 3%       ~     (p=0.942 n=3+3)
WindowFunctions/(func:rank,_ndv:1000,_rows:100000)-12            6.79ms ± 0%    9.61ms ± 1%    +41.60%  (p=0.000 n=3+3)
WindowFunctions/(func:dense_rank,_ndv:1000,_rows:100000)-12      6.81ms ± 1%    9.70ms ± 1%    +42.59%  (p=0.000 n=3+3)
WindowFunctions/(func:cume_dist,_ndv:1000,_rows:100000)-12       6.82ms ± 3%   10.09ms ± 1%    +47.99%  (p=0.000 n=3+3)
WindowFunctions/(func:percent_rank,_ndv:1000,_rows:100000)-12    6.86ms ± 1%    9.62ms ± 0%    +40.25%  (p=0.000 n=3+3)
WindowFunctions/(func:ntile,_ndv:1000,_rows:100000)-12           6.31ms ± 1%    6.17ms ± 1%       ~     (p=0.050 n=3+3)
WindowFunctions/(func:lead,_ndv:1000,_rows:100000)-12            8.04ms ± 1%   12.30ms ± 0%    +52.91%  (p=0.000 n=3+3)
WindowFunctions/(func:first_value,_ndv:1000,_rows:100000)-12     6.28ms ± 1%    6.44ms ± 0%       ~     (p=0.063 n=3+3)
WindowFunctions/(func:last_value,_ndv:1000,_rows:100000)-12      6.36ms ± 0%    6.40ms ± 1%       ~     (p=0.415 n=3+3)
WindowFunctions/(func:nth_value,_ndv:1000,_rows:100000)-12       6.48ms ± 1%    6.47ms ± 1%       ~     (p=0.871 n=3+3)

name                                                           old alloc/op   new alloc/op   delta
WindowRows/(func:row_number,_ndv:10,_rows:1000)-12               44.7kB ± 0%    44.7kB ± 0%       ~     (zero variance)
WindowRows/(func:row_number,_ndv:1000,_rows:1000)-12             81.0kB ± 1%    80.5kB ± 1%       ~     (p=0.523 n=3+3)
WindowRows/(func:row_number,_ndv:10,_rows:100000)-12              263kB ± 0%     263kB ± 0%       ~     (p=0.443 n=3+3)
WindowRows/(func:row_number,_ndv:1000,_rows:100000)-12            301kB ± 0%     299kB ± 1%       ~     (p=0.423 n=3+3)
WindowFunctions/(func:row_number,_ndv:1000,_rows:100000)-12       301kB ± 0%     301kB ± 0%       ~     (zero variance)
WindowFunctions/(func:rank,_ndv:1000,_rows:100000)-12             307kB ± 0%    6917kB ± 0%  +2155.27%  (p=0.000 n=3+3)
WindowFunctions/(func:dense_rank,_ndv:1000,_rows:100000)-12       306kB ± 0%    6921kB ± 0%  +2160.67%  (p=0.000 n=3+3)
WindowFunctions/(func:cume_dist,_ndv:1000,_rows:100000)-12        306kB ± 0%    6923kB ± 0%  +2161.01%  (p=0.000 n=3+3)
WindowFunctions/(func:percent_rank,_ndv:1000,_rows:100000)-12     306kB ± 0%    6929kB ± 0%  +2162.99%  (p=0.000 n=3+3)
WindowFunctions/(func:ntile,_ndv:1000,_rows:100000)-12            307kB ± 0%     301kB ± 0%     -1.95%  (p=0.000 n=3+3)
WindowFunctions/(func:lead,_ndv:1000,_rows:100000)-12             306kB ± 0%    5984kB ± 0%  +1857.22%  (p=0.000 n=3+3)
WindowFunctions/(func:first_value,_ndv:1000,_rows:100000)-12      301kB ± 0%     301kB ± 0%       ~     (zero variance)
WindowFunctions/(func:last_value,_ndv:1000,_rows:100000)-12       301kB ± 0%     301kB ± 0%       ~     (zero variance)
WindowFunctions/(func:nth_value,_ndv:1000,_rows:100000)-12        301kB ± 0%     301kB ± 0%       ~     (zero variance)

name                                                           old allocs/op  new allocs/op  delta
WindowRows/(func:row_number,_ndv:10,_rows:1000)-12                 74.0 ± 0%      74.0 ± 0%       ~     (zero variance)
WindowRows/(func:row_number,_ndv:1000,_rows:1000)-12                698 ± 3%       690 ± 1%       ~     (p=0.523 n=3+3)
WindowRows/(func:row_number,_ndv:10,_rows:100000)-12              1.94k ± 0%     1.94k ± 0%       ~     (p=1.000 n=3+3)
WindowRows/(func:row_number,_ndv:1000,_rows:100000)-12            3.00k ± 0%     3.00k ± 0%       ~     (p=0.423 n=3+3)
WindowFunctions/(func:row_number,_ndv:1000,_rows:100000)-12       3.00k ± 0%     3.00k ± 0%       ~     (zero variance)
WindowFunctions/(func:rank,_ndv:1000,_rows:100000)-12             3.01k ± 0%    23.81k ± 0%   +691.91%  (p=0.000 n=3+3)
WindowFunctions/(func:dense_rank,_ndv:1000,_rows:100000)-12       3.01k ± 0%    23.83k ± 0%   +692.33%  (p=0.000 n=3+3)
WindowFunctions/(func:cume_dist,_ndv:1000,_rows:100000)-12        3.01k ± 0%    23.83k ± 0%   +692.54%  (p=0.000 n=3+3)
WindowFunctions/(func:percent_rank,_ndv:1000,_rows:100000)-12     3.01k ± 0%    23.85k ± 0%   +693.17%  (p=0.000 n=3+3)
WindowFunctions/(func:ntile,_ndv:1000,_rows:100000)-12            3.01k ± 0%     3.00k ± 0%       ~     (zero variance)
WindowFunctions/(func:lead,_ndv:1000,_rows:100000)-12             3.01k ± 0%   111.01k ± 0%  +3591.58%  (p=0.000 n=3+3)
WindowFunctions/(func:first_value,_ndv:1000,_rows:100000)-12      3.00k ± 0%     3.00k ± 0%       ~     (zero variance)
WindowFunctions/(func:last_value,_ndv:1000,_rows:100000)-12       3.00k ± 0%     3.00k ± 0%       ~     (zero variance)
WindowFunctions/(func:nth_value,_ndv:1000,_rows:100000)-12        3.00k ± 0%     3.00k ± 0%       ~     (zero variance)
```

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by SunRunAway** [comment]
The unit test fails because it relies on #11673 
Please ignore it and merge this pr firstly, then merge #11673

**Comment by foreyes** [comment]
Did you reference a wrong PR number? @SunRunAway

**Comment by SunRunAway** [comment]
@foreyes sorry, fixed.

File: executor/aggfuncs/func_cume_dist.go [hunk 1]
```diff
@@ -42,14 +43,28 @@ func (r *cumeDist) ResetPartialResult(pr PartialResult) {
 
 func (r *cumeDist) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) error {
 	p := (*partialResult4CumeDist)(pr)
-	p.rows = append(p.rows, rowsInGroup...)
+	for _, row := range rowsInGroup {
+		rowDatum := make([]*types.Datum, len(r.colIdx))
+		for i, idx := range r.colIdx {
+			d := row.GetDatum(idx, r.retTypes[i])
```
Comment (lamxTyler) [inline executor/aggfuncs/func_cume_dist.go]: If the datum type is string, I think it is still an reference.

Comment (SunRunAway) [reply executor/aggfuncs/func_cume_dist.go]: PTAL again.

File: executor/aggfuncs/func_cume_dist.go [hunk 2]
```diff
@@ -26,7 +27,7 @@ type cumeDist struct {
 type partialResult4CumeDist struct {
 	curIdx   int
 	lastRank int
-	rows     []chunk.Row
+	rows     [][]*types.Datum
```
Comment (zz-jason) [inline executor/aggfuncs/func_cume_dist.go]: It's better to use a `Chunk` to buffer the underlying data, and still use `[]chunk.Row` to store the pointers.

File: executor/aggfuncs/func_rank.go [hunk 3]
```diff
@@ -28,7 +29,7 @@ type rank struct {
 type partialResult4Rank struct {
 	curIdx   int64
 	lastRank int64
-	rows     []chunk.Row
+	rows     [][]*types.Datum
```
Comment (zz-jason) [inline executor/aggfuncs/func_rank.go]: ditto

File: executor/aggfuncs/func_cume_dist.go (hunk 2 above)
Comment (SunRunAway) [reply executor/aggfuncs/func_cume_dist.go]: It's hard to copy some columns of a Row to a new Row/Chunk.
Vectorizing window functions is a good opportunity to optimize this, maybe later.

File: executor/aggfuncs/func_lead_lag.go [hunk 4]
```diff
@@ -14,76 +14,196 @@
...
 }
 
 type lead struct {
 	baseLeadLag
 }
 
+func (v *lead) AllocPartialResult() PartialResult {
+	return PartialResult(&partialResult4Lead{
+		defaultExtractors: circleBuf{
+			buf:  make([]valueExtractor, 0, mathutil.MinUint64(v.offset, 1000)),
```
Comment (lamxTyler) [inline executor/aggfuncs/func_lead_lag.go]: Define a constant for `1000`.

File: executor/aggfuncs/func_lead_lag.go [hunk 5]
```diff
@@ -14,76 +14,196 @@
...
+			e := buildValueExtractor(v.retTp)
+			err = e.extractRow(sctx, v.args[0], row)
+			if err != nil {
+				return err
+			}
+			p.extractors = append(p.extractors, e)
+		}
+		if v.offset > 0 {
+			if !v.defaultExpr.ConstItem() {
+				// we must cache the results of last v.offset lines
```
Comment (lamxTyler) [inline executor/aggfuncs/func_lead_lag.go]: `We ... lines.`, capitalize the first character and add dot at the end of line.

File: executor/aggfuncs/func_lead_lag.go [hunk 6]
```diff
@@ -14,76 +14,200 @@
...
+	defaultExtractors []valueExtractor
+}
+
 type lag struct {
 	baseLeadLag
 }
 
+func (v *lag) AllocPartialResult() PartialResult {
+	return PartialResult(&partialResult4Lag{
+		defaultExtractors: make([]valueExtractor, 0, mathutil.MinUint64(v.offset, 1000)),
```
Comment (lamxTyler) [inline executor/aggfuncs/func_lead_lag.go]: Ditto, use the const instead.

File: executor/aggfuncs/func_lead_lag_test.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
+// Copyright 2018 PingCAP, Inc.
```
Comment (lamxTyler) [inline executor/aggfuncs/func_lead_lag_test.go]: 2019

File: types/mydecimal.go [hunk 8]
```diff
@@ -250,6 +250,21 @@ func (d *MyDecimal) GetDigitsFrac() int8 {
 	return d.digitsFrac
 }
 
+// Copy copies a new *MyDecimal from itself.
+func (d *MyDecimal) Copy() *MyDecimal {
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply types/mydecimal.go]: `BinaryJSON` uses `Copy`, so I use the same for a unifying name.
https://godoc.org/github.com/pingcap/tidb/types/json#BinaryJSON.Copy

File: executor/aggfuncs/func_rank.go [hunk 9]
```diff
@@ -28,7 +28,44 @@ type rank struct {
 type partialResult4Rank struct {
 	curIdx   int64
 	lastRank int64
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply executor/aggfuncs/func_rank.go]: It is used in `partialResult4Rank.updatePartialResult`.

File: session/tidb.go [hunk 10]
```diff
@@ -290,10 +288,10 @@ func GetRows4Test(ctx context.Context, sctx sessionctx.Context, rs sqlexec.Recor
 			break
 		}
 
+		iter := chunk.NewIterator4Chunk(req.CopyConstruct())
 		for row := iter.Begin(); row != iter.End(); row = iter.Next() {
 			rows = append(rows, row)
 		}
-		req = chunk.Renew(req, sctx.GetSessionVars().MaxChunkSize)
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply session/tidb.go]: See the comment above,
> Must reuse `req` for imitating server.(*clientConn).writeChunks

File: executor/aggfuncs/func_lead_lag.go [hunk 11]
```diff
@@ -14,76 +14,200 @@
...
-func (v *baseLeadLag) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) error {
-	p := (*partialResult4LeadLag)(pr)
-	p.rows = append(p.rows, rowsInGroup...)
-	return nil
+func (cb *circleBuf) get() (e valueExtractor) {
+	if len(cb.buf) < cb.size {
+		e = cb.buf[cb.head]
+		cb.head++
+	} else {
+		if cb.tail >= cb.size {
```
Comment (zz-jason) [inline executor/aggfuncs/func_lead_lag.go]: should use `head`?

File: executor/aggfuncs/func_lead_lag.go [hunk 12]
```diff
@@ -14,76 +14,200 @@
...
+func (cb *circleBuf) reset() {
+	cb.buf = cb.buf[:0]
+	cb.head, cb.tail = 0, 0
 }
 
-func (v *baseLeadLag) ResetPartialResult(pr PartialResult) {
-	p := (*partialResult4LeadLag)(pr)
-	p.rows = p.rows[:0]
-	p.curIdx = 0
+func (cb *circleBuf) append(e valueExtractor) {
```
Comment (zz-jason) [inline executor/aggfuncs/func_lead_lag.go]: should we expand the buffer size when it's full?

**Pull request merged**